### PR TITLE
Faster, less memory by caching dpo logprobs

### PR DESCRIPTION
**Pull request by vwxyzjn**
This PR caches the DPO logprobs to save memory (see [report](https://wandb.ai/ai2-llm/open_instruct_internal/reports/DPO-cache-logprobs--Vmlldzo5ODM5Mjkw))

image

I put this as a separate `dpo_tune_cach.py`. We can prob the  `dpo_tune.py` with this implementation, but maybe there are some edge cases like dataloader resuming with dpo cache logprobs or lora stuff to test, so I am also ok with just a `dpo_tune_cache.py` at the moment until we are done with the release. WDYT?

### CODE REVIEW

File: open_instruct/dpo_tune_cache.py [hunk 1]
```diff
@@ -0,0 +1,1169 @@
...
+    progress_bar.update(completed_steps)
+
+    local_metrics = torch.zeros((20), device=accelerator.device)
+    episode = 0
+    for epoch in range(starting_epoch, args.num_train_epochs):
+        model.train()
+        train_dataloader.set_epoch(epoch)
+        if last_checkpoint_path and resume_step is not None:
+            # We skip the first `n` batches in the dataloader when resuming from a checkpoint
+            active_dataloader = accelerator.skip_first_batches(train_dataloader, resume_step)
```
Comment (hamishivi) [inline open_instruct/dpo_tune_cache.py:957]: If we skip initial steps, and then resume, I don't think `step` gets updated, but we need that to reference the cached logprobs correctly?

File: open_instruct/dpo_tune_cache.py [hunk 2]
```diff
@@ -0,0 +1,1169 @@
...
+            resume_step -= starting_epoch * len(train_dataloader)
+
+    print(f"Starting from epoch {starting_epoch} and step {completed_steps}.")
+
+    # Cache the logprobs
+    average_log_prob_loss_types = ["simpo", "dpo_norm"]
+    average_log_prob = args.dpo_loss_type in average_log_prob_loss_types
+    if args.dpo_loss_type == "dpo" or args.dpo_loss_type == "dpo_norm":
+        cached_reference_chosen_logps, cached_reference_rejected_logps = get_cache_ref_logprobs(
+            model, train_dataloader, accelerator, average_log_prob
```
Comment (hamishivi) [inline open_instruct/dpo_tune_cache.py]: Did  you double check that the order of the samples here matches the order of the samples later? Doesn't the dataloader reshuffle every epoch?

**Review by hamishivi — COMMENTED** [review]
Left some comments... the curves look good but i want to be really certain that this is not changing things.

File: open_instruct/dpo_tune_cache.py (hunk 2 above)
Comment (vwxyzjn) [reply open_instruct/dpo_tune_cache.py]: image

Yeah, the order of the samples changes. I added logic to handle the epochs stuff.

image

File: open_instruct/dpo_tune_cache.py (hunk 1 above)
Comment (vwxyzjn) [reply open_instruct/dpo_tune_cache.py:957]: Added the corresponding logic to the PR.

**Comment by vwxyzjn** [comment]
I have added the logic to handle those two scenarios. @hamishivi WDYT about merging this version as is (not replacing the existing `dpo_tune.py`) and let people test it out more extensively before merging?

File: open_instruct/dpo_tune_cache.py (hunk 2 above)
Comment (hamishivi) [reply open_instruct/dpo_tune_cache.py]: Right, but in a multi-epoch setting don't we need to change how we index into the cached logprobs to match the new ordering when we swap into a new epoch? Our configs now use 1 epoch, but it'd be good to support multi-epoch.

**Comment by hamishivi** [comment]
@vwxyzjn Yeah, happy to keep this separate from dpo_tune for now in the interest of time. I think its fine to let people test out, it might be good to test this on a few datasets and hyperparameter settings against the original dpo script to be safe that the caching logic is fine. I've been bit on adding caching logic and then it actually being wrong before, hence my caution 😅

**Review by hamishivi — APPROVED** [review]

**Comment by vwxyzjn** [comment]
Of course @hamishivi! Thanks for checking closely. Also tested it with 3 epochs :) 

image
